src/gui/properties/speedplotview.cpp

```diff
@@ -118,55 +114,71 @@ namespace
     }
 }
 
-SpeedPlotView::Averager::Averager(int divider, boost::circular_buffer<PointData> &sink)
-    : m_divider(divider)
-    , m_sink(sink)
-    , m_counter(0)
-    , m_accumulator {}
+SpeedPlotView::Averager::Averager(const milliseconds duration, const milliseconds resolution)
+    : m_resolution {resolution}
+    , m_maxDuration {duration}
 {
+    m_timeSinceLastPush.start();
 }
 
-void SpeedPlotView::Averager::push(const PointData &pointData)
+bool SpeedPlotView::Averager::push(const SampleData &sampleData)
 {
     // Accumulator overflow will be hit in worst case on longest used averaging span,
     // defined by divider value. Maximum divider is DIVIDER_24HOUR = 144
     // Using int32 for accumulator we get overflow when transfer speed reaches 2^31/144 ~~ 14.2 MBytes/s.
     // With quint64 this speed limit is 2^64/144 ~~ 114 PBytes/s.
     // This speed is inaccessible to an ordinary user.
-    m_accumulator.x += pointData.x;
+    ++m_counter;
     for (int id = UP; id < NB_GRAPHS; ++id)
-        m_accumulator.y[id] += pointData.y[id];
-    m_counter = (m_counter + 1) % m_divider;
-    if (m_counter != 0)
-        return; // still accumulating
+        m_accumulator[id] += sampleData[id];
+
+    // system may go to sleep, that can cause very big elapsed interval
+    const milliseconds maxElapsed {static_cast<qint64>(std::ceil(BitTorrent::Session::instance()->refreshInterval() * 1.25))};
+    const milliseconds elapsed {qMin(milliseconds {m_timeSinceLastPush.elapsed()}, maxElapsed)};
+    if (elapsed < m_resolution)
+        return false; // still accumulating
+
     // it is time final averaging calculations
+    m_currentDuration += elapsed;
+    if ((m_currentDuration > m_maxDuration)
+        // once we go above the max duration never go below that, otherwise it will cause empty space in graphs
+        && (m_currentDuration - (!m_sink.empty() ? m_sink.front().duration : milliseconds {0}) > m_maxDuration))
+    {
+        m_sink.set_capacity(m_sink.size());
+        if (!m_sink.empty())
+            m_currentDuration -= m_sink.front().duration;
+    }
+    else if (m_sink.size() + 1 >= m_sink.capacity())
```

Comment:
Let's start small...

Where did this condition come from?
If I am not mistaken, then you increase the capacity in two cases, either when the size is less by one than the current capacity, or when they are equal (since the size cannot be greater than the capacity). So why not just leave the only condition (size equals capacity), as in other case (the size is less than capacity) you can safely call `push_back()` (later in the code), without fear of losing the first item of queue (this is the condition for protection from loss of the first item of queue, am I right?).

Reply:
yes this condition can be `m_sink.size() == m_sink.capacity()`